layers: Use a concurrent_unordered_map for object tracking in dispatch

jeffbolznv opened the pull request:
Add a vl_concurrent_unordered_map class that acts like an unordered_map but
internally synchronizes concurrent accesses, and split the locks into 2^N
buckets. Remove the separate dispatch_lock which is no longer needed.

Replaces #1137. I'm seeing a meaningful improvement in performance relative to that PR. I think the bulk of the benefit is coming from reduced lock contention caused by frequent AllocateDescriptorSets calls which were taking the write lock.

mark-lunarg commented:
Running through CI.

jzulauf-lunarg commented on layers/generated/layer_chassis_dispatch.cpp:
```diff
@@ -200,14 +195,12 @@ VkResult DispatchCreateGraphicsPipelines(VkDevice device, VkPipelineCache pipeli
-                const auto subpasses_uses_it = layer_data->renderpasses_states.find(layer_data->Unwrap(pCreateInfos[idx0].renderPass));
-                if (subpasses_uses_it != layer_data->renderpasses_states.end()) {
-                    const auto &subpasses_uses = subpasses_uses_it->second;
+                if (layer_data->renderpasses_states.contains(layer_data->Unwrap(pCreateInfos[idx0].renderPass))) {
```
This double your hash overhead, no?

jzulauf-lunarg commented on layers/vk_layer_utils.h:
```diff
@@ -159,3 +159,66 @@ static inline int u_ffs(int val) {
+        return maps[h].erase(key);
+    }
+
+    bool contains(const Key &key) {
```
to avoid the double hash/lock of the contains/at pattern, perhaps some customer accessor that returns a T *, with null denoting not found (const and non const flavors) would yield an additional performance benefit.

jeffbolznv replied:
The contains/at were rare enough in this change that I didn't bother with this, but I have a followon change (to use this class for the object tracker) where I did this. Rather than a T*, I have a "find_result" type that acts vaguely like an iterator and lets you do roughly:

```
   auto iter = map.find(key);
   if (iter != map.end()) {
        auto element = iter->second;
```

This lets a lot of code use this and not require any modifications (or very small modifications). I'll pull that change into this branch in the morning.

jeffbolznv replied:
Pushed the update with this improvement

jzulauf-lunarg commented on scripts/layer_chassis_generator.py:
```diff
@@ -346,27 +347,28 @@ class ValidationObject {
-            return (HandleType)unique_id_mapping[reinterpret_cast<uint64_t const &>(wrappedHandle)];
+            if (wrappedHandle == 0)
+                return (HandleType)0;
+            return (HandleType)unique_id_mapping.at(reinterpret_cast<uint64_t const &>(wrappedHandle));
```
at throws on unknown wrappedHandle (unlike [] in previous).  Using a `find` -> `find_result` would avoid

jeffbolznv replied:
I was assuming a well-behaved app, but that's not a valid assumption in this code. I agree, a find would be better.

jzulauf-lunarg commented on layers/generated/chassis.h:
```diff
@@ -2467,27 +2468,28 @@ class ValidationObject {
+            return (HandleType)unique_id_mapping.at(reinterpret_cast<uint64_t const &>(wrappedHandle));
         }
 
         // Wrap a newly created handle with a new unique ID, and return the new ID -- must hold lock.
```
Comment out of date?

jzulauf-lunarg commented on scripts/layer_chassis_generator.py:
```diff
@@ -346,27 +347,28 @@ class ValidationObject {
+        vl_concurrent_unordered_map<VkDescriptorPool, std::unordered_set<VkDescriptorSet>> pool_descriptor_sets_map;
 
 
         // Unwrap a handle.  Must hold lock.
```
Comment out of date

jzulauf-lunarg commented on layers/vk_layer_utils.h:
```diff
@@ -159,3 +159,127 @@ static inline int u_ffs(int val) {
+// for existing code that combines the existence check and lookup in a single
+// operation (and thus a single lock). i.e.:
+//
+//      auto iter = map.find(key);
```
The find_result will race against erase, unless the find_result holds the lock.  Const find_result could hold a read_lock and non-const find_results a write_lock.

jzulauf-lunarg commented on layers/vk_layer_utils.h:
```diff
@@ -159,3 +159,127 @@ static inline int u_ffs(int val) {
+template <typename Key, typename T, int BUCKETSLOG2 = 2>
+class vl_concurrent_unordered_map {
+   public:
+    T &at(const Key &key) {
```
anytime this is used on the RHS you create a race vs. erase.  Need some return that moves the read_lock_guard_t to the scope of the caller.

jzulauf-lunarg commented on layers/vk_layer_utils.h:
```diff
@@ -159,3 +159,127 @@ static inline int u_ffs(int val) {
+
+    // Note: operator[] takes a write lock. For reads, prefer at() which only
+    // takes a read lock.
+    T &operator[](const Key &key) {
```
same comment re: RHS usage as `at`

jeffbolznv commented on layers/vk_layer_utils.h:
```diff
@@ -159,3 +159,127 @@ static inline int u_ffs(int val) {
+        bool operator!=(const find_result &other) const { return !(*this == other); }
+
+        // Make -> act kind of like an iterator.
+        std::pair<bool, T &> *operator->() { return &result; }
```
If I used a pointer type, then a bunch of calling code would have to change. Part of the purpose of using this pair was to make it act like an iterator and allow iter->second to return the contained value.

jeffbolznv commented on layers/generated/layer_chassis_dispatch.cpp:
```diff
@@ -200,7 +195,6 @@ VkResult DispatchCreateGraphicsPipelines(VkDevice device, VkPipelineCache pipeli
     safe_VkGraphicsPipelineCreateInfo *local_pCreateInfos = nullptr;
     if (pCreateInfos) {
         local_pCreateInfos = new safe_VkGraphicsPipelineCreateInfo[createInfoCount];
-        std::lock_guard<std::mutex> lock(dispatch_lock);
```
Hmm, maybe it was a mistake to return a reference, and I should return the value instead. I originally implemented this class just for the unordered_map<uint64_t,uint64_t> where returning a value would be fine, and then later expanded it to the other types (returning by value may be fine for some of these types, even if it's expensive to make a copy of the value, since they're used rarely).

I don't want anything outside the class to hold any of its internal locks, it introduces the possibility of deadlock (e.g. thread 0 holds read lock 0 and tries to take write lock 1, thread 1 holds read lock 1 and tries to take write lock 0).

jzulauf-lunarg replied:
The `find_result` could certainly store the T by value and be safe enough for this specific case.  The would work for shared_ptr's as well quite nicely, which is another case we will care about RSN.

We still need to guard the "other" map in addition to avoiding the race.

As for "order of locks" it's something we're going to need to be aware of eventually I think, and it's just part of design discipline. If at any point, any thread is going to need A and B locked, we just need to document the the ordering required. 

I've even seen mutex schemes that assign unique "levels" to mutex's s.t. debug code can be added to assert that ordering is always maintained (e.g. a thread local of the deepest "level" locked, and asserting when a lock of a shallower level is attempted...)

jeffbolznv commented:
I've pushed another commit that changes the concurrent_unordered_map to return by value, and reverts some of the more complex maps to use a mutex. I *think* this addresses your concerns, but it's possible I've missed something.

jzulauf-lunarg commented on scripts/layer_chassis_dispatch_generator.py:
```diff
@@ -293,10 +288,17 @@ class LayerChassisDispatchOutputGenerator(OutputGenerator):
     auto layer_data = GetLayerDataPtr(get_dispatch_key(device), layer_data_map);
     if (!wrap_handles) return layer_data->device_dispatch_table.DestroyRenderPass(device, renderPass, pAllocator);
-    std::unique_lock<std::mutex> lock(dispatch_lock);
+    write_dispatch_lock_guard_t lock(dispatch_lock);
```
Does this need to be held before the dispatch?

jeffbolznv replied:
I don't think so, I'll move it after.

jzulauf-lunarg commented on scripts/layer_chassis_dispatch_generator.py:
```diff
@@ -401,8 +398,15 @@ class LayerChassisDispatchOutputGenerator(OutputGenerator):
+    } else {
+        swapchain = (VkSwapchainKHR)0;
+    }
+
```
Can the unlock be moved up to 401?

jzulauf-lunarg commented on scripts/layer_chassis_dispatch_generator.py:
```diff
@@ -452,8 +455,15 @@ class LayerChassisDispatchOutputGenerator(OutputGenerator):
+    } else {
+        descriptorPool = (VkDescriptorPool)0;
+    }
+
```
Earlier unlock here?

jzulauf-lunarg commented on scripts/layer_chassis_dispatch_generator.py:
```diff
@@ -621,11 +626,18 @@ class LayerChassisDispatchOutputGenerator(OutputGenerator):
+    } else {
+        descriptorUpdateTemplate = (VkDescriptorUpdateTemplate)0;
+    }
+
```
Earlier unlock?

jzulauf-lunarg commented on layers/vk_layer_utils.h:
```diff
@@ -159,3 +159,118 @@ static inline int u_ffs(int val) {
+    struct {
+        lock_t lock;
+        // Put each lock on its own cache line to avoid false cache line sharing.
+        char padding[64];
```
would an alignment pragma here be too much of a pain to do crossing platform? or padding[CL_SIZE - sizeof(lock_t))?

jeffbolznv replied:
I'll change to 64-sizeof(lock_t)

jeffbolznv commented on scripts/layer_chassis_generator.py:
```diff
@@ -346,7 +347,7 @@ class ValidationObject {
         // Handle Wrapping Data
         // Reverse map display handles
-        std::unordered_map<VkDisplayKHR, uint64_t> display_id_reverse_mapping;
+        vl_concurrent_unordered_map<VkDisplayKHR, uint64_t> display_id_reverse_mapping;
```
VkDisplay usage should be very rare, so I'll just decrease this to one bucket.

jzulauf-lunarg replied:
After this lands, we should do some profiling, but I think you're right.  It would be nice to figure out how to get the reader/writer lock working in Linux/g++, too after.

jeffbolznv commented on scripts/layer_chassis_generator.py:
```diff
@@ -362,31 +363,33 @@ class ValidationObject {
-            // TODO: don't use operator[] here.
-            return (HandleType)unique_id_mapping[reinterpret_cast<uint64_t const &>(wrappedHandle)];
+            auto iter = unique_id_mapping.find(reinterpret_cast<uint64_t const &>(wrappedHandle));
+            if (iter == unique_id_mapping.end())
```
I like the idea of matching STL style as close as possible in case we ever swap this out for another class, e.g. TBB's concurrent_unordered_map.

jzulauf-lunarg replied:
Fair enough.  If this ever shows up in profiling I'll be shocked, in any case. Stet.

jeffbolznv commented on scripts/layer_chassis_generator.py:
```diff
@@ -225,8 +226,8 @@ class LayerChassisOutputGenerator(OutputGenerator):
-extern uint64_t global_unique_id;
-extern std::unordered_map<uint64_t, uint64_t> unique_id_mapping;
+extern std::atomic<uint64_t> global_unique_id;
+extern vl_concurrent_unordered_map<uint64_t, uint64_t, 8> unique_id_mapping;
```
I think I just set it to something large and never revisited it. On one hand, there's a pretty high penalty to a collision since the OS will likely unschedule the thread, but on the other hand this map is dominated by reads with relatively few writes, so few collisions. I'm going to change it to bucketslog2=4 for now, which is probably plenty.

jzulauf-lunarg replied:
4 or 6 (like you did in thread_safety) is probably a reasonable number.  As noted above, after this lands some profiling and tuning would be reasonable.

jeffbolznv commented on layers/generated/layer_chassis_dispatch.cpp:
```diff
@@ -302,10 +308,17 @@ VkResult DispatchCreateRenderPass2KHR(VkDevice device, const VkRenderPassCreateI
+        renderPass = (VkRenderPass)iter->second;
+        unique_id_mapping.erase(renderPass_id);
+    } else {
+        renderPass = (VkRenderPass)0;
```
I think it's currently undefined/uninitialized. I could potentially add an initializer to the class, but it seems odd to me to rely on the value from the "iterator" when it wasn't found.

jeffbolznv commented on layers/generated/layer_chassis_dispatch.cpp:
```diff
@@ -302,10 +308,17 @@ VkResult DispatchCreateRenderPass2KHR(VkDevice device, const VkRenderPassCreateI
-    renderPass = (VkRenderPass)unique_id_mapping[renderPass_id];
-    unique_id_mapping.erase(renderPass_id);
+
+    auto iter = unique_id_mapping.find(renderPass_id);
```
I'm torn. The pop does sound elegant, but staying close to STL is also desirable...

I'll sleep on it. I think I've addressed the rest of your feedback and will push an update. Thanks for the great reviews!

jeffbolznv replied:
OK, I've added pop and squashed the commits.

mark-lunarg commented:
Rebased and pushed to Master.
